### PR TITLE
scene: add wlr_scene_buffer

### PR DESCRIPTION
**Pull request by emersion**
This new scene-graph node displays a wlr_buffer.

Closes: https://github.com/swaywm/wlroots/issues/3161

### CODE REVIEW

**Comment by emersion** [comment]
To make this usable for rendering saved buffers we also need to add a few more functions for scaling/cropping.

**Comment by emersion** [comment]
Pushed two more commits to add scaling and cropping.

**Comment by emersion** [comment]
One thing worth noting is that it should be possible to implement `wlr_scene_surface` on top of `wlr_scene_buffer`. Maybe we should do that at some point. Need to investigate how that would impact surface-specific integrations with the scene-graph, such as auto `wl_surface.enter`/`leave`.

File: types/scene/wlr_scene.c [hunk 1]
```diff
@@ -220,6 +232,119 @@ void wlr_scene_rect_set_color(struct wlr_scene_rect *rect, const float color[sta
...
+		return client_buffer->texture;
+	}
+
+	if (scene_buffer->texture != NULL) {
+		return scene_buffer->texture;
+	}
+
+	scene_buffer->texture =
+		wlr_texture_from_buffer(renderer, scene_buffer->buffer);
+	if (scene_buffer->texture == NULL) {
```
Comment (vyivel) [inline types/scene/wlr_scene.c:304]: Is this check really necessary?

Comment (emersion) [reply types/scene/wlr_scene.c:304]: Yes -- the name might not be well-chosen but `wlr_texture_from_buffer` will import/upload the buffer. This can fail.

Comment (vyivel) [reply types/scene/wlr_scene.c:304]: Ah, sorry, that's not what I meant. Returning NULL if the value is NULL is redundant if said value is returned otherwise.

Comment (emersion) [reply types/scene/wlr_scene.c]: Oh, right, fixed!

**Comment by emersion** [comment]
> I think a struct wlr_scene_node_impl would be a nice addition in the future to clear the big switches here and there.

This has been discussed here: https://github.com/swaywm/wlroots/pull/3140#issuecomment-907100708

tl;dr not sure it'd simplify things.

**Comment by bl4ckb0ne** [comment]
Good point, it does make sense to keep the impl indeed.

Don't forget to rebase the patchset!
